File: frontend/src/utils/scorer.js
export function scorePrompt(text) {
  if (!text) return { clarity: 0, specificity: 0, context: 0, structure: 0, total: 0, grade: "F" }
  
  const words = text.split(/\s+/).filter(Boolean)
  const wc = words.length
  const tl = text.toLowerCase()

  const vague = ["thing","stuff","something","somehow","whatever",
                 "maybe","perhaps","kind of","sort of","etc"]
  const action = ["build","create","return","validate","generate",
                  "implement","write","define","configure","deploy"]
  const vagueC  = vague.filter(w => tl.includes(w)).length
  const actionC = action.filter(w => tl.includes(w)).length
  const clarity = Math.max(0, Math.min(25, 15 - (vagueC*3) + (actionC*2)))

  const tech = ["api","json","function","class","database","endpoint",
                "schema","component","async","react","python","sql"]
  const techC = tech.filter(t => tl.includes(t)).length
  const lenBonus = Math.min(10, Math.floor(wc / 5))
  const specificity = Math.max(0, Math.min(25, (techC*3) + lenBonus))

  const ctxWords = ["because","in order to","given that","using",
                    "with","for the purpose","so that","when"]
  const ctxC = ctxWords.filter(c => tl.includes(c)).length
  const context = Math.max(0, Math.min(25, ctxC*5 + (wc > 30 ? 5 : 0)))

  const hasNumbers = /\d+[\.\)]\s/.test(text)
  const hasBullets = /[-*•]\s/.test(text)
  const hasRole    = ["you are","act as","as a"].some(r => tl.includes(r))
  const hasFormat  = ["return","output","format","respond with"].some(f => tl.includes(f))
  const structure  = Math.min(25,
    (hasNumbers?7:0) + (hasBullets?6:0) + (hasRole?6:0) + (hasFormat?6:0))

  const total = clarity + specificity + context + structure
  const grade = total>=85?"A": total>=70?"B": total>=55?"C": total>=40?"D":"F"
  
  return { clarity, specificity, context, structure, total, grade }
}

File: backend/app/services/scorer.py
import re

def score_prompt(text: str) -> dict:
    words = text.split()
    word_count = len(words)
    tl = text.lower()

    vague = ["thing","stuff","something","somehow","whatever",
             "maybe","perhaps","kind of","sort of","etc"]
    action = ["build","create","return","validate","generate",
              "implement","write","define","configure","deploy"]
    vague_c  = sum(1 for w in vague  if w in tl)
    action_c = sum(1 for w in action if w in tl)
    clarity = max(0, min(25, 15 - (vague_c * 3) + (action_c * 2)))

    tech = ["api","json","function","class","database","endpoint",
            "schema","component","async","react","python","sql"]
    tech_c = sum(1 for t in tech if t in tl)
    len_bonus = min(10, word_count // 5)
    specificity = max(0, min(25, (tech_c * 3) + len_bonus))

    ctx_words = ["because","in order to","given that","using",
                 "with","for the purpose","so that","when"]
    ctx_c = sum(1 for c in ctx_words if c in tl)
    context = max(0, min(25, ctx_c * 5 + (5 if word_count > 30 else 0)))

    has_numbers = bool(re.search(r'\d+[\.\)]\s', text))
    has_bullets = bool(re.search(r'[-*•]\s', text))
    has_role    = any(r in tl for r in ["you are","act as","as a"])
    has_format  = any(f in tl for f in ["return","output","format","respond with"])
    structure = min(25,
        (7 if has_numbers else 0) + (6 if has_bullets else 0) +
        (6 if has_role else 0)    + (6 if has_format else 0))

    total = clarity + specificity + context + structure
    grade = ("A" if total >= 85 else "B" if total >= 70 else
             "C" if total >= 55 else "D" if total >= 40 else "F")
    return {
        "clarity": clarity, "specificity": specificity,
        "context": context, "structure": structure,
        "total": total, "grade": grade
    }
